cpp/include/raft/sparse/mst/detail/mst_solver_inl.cuh

```diff
@@ -0,0 +1,400 @@
+  weight_t max = alteration_max();
+
+  // pool of rand values
+  rmm::device_vector<weight_t> rand_values(v);
+
+  // Random number generator
+  curandGenerator_t randGen;
+  curandCreateGenerator(&randGen, CURAND_RNG_PSEUDO_DEFAULT);
+  curandSetPseudoRandomGeneratorSeed(randGen, 1234567);
+
+  // Initialize rand values
+  auto curand_status = curand_generate_uniformX(
+    randGen, thrust::raw_pointer_cast(rand_values.data()), v);
+  RAFT_EXPECTS(curand_status == CURAND_STATUS_SUCCESS, "MST: CURAND failed");
+  curand_status = curandDestroyGenerator(randGen);
+  RAFT_EXPECTS(curand_status == CURAND_STATUS_SUCCESS,
+               "MST: CURAND cleanup failed");
+
+  //Alterate the weights, make all undirected edge weight unique while keeping Wuv == Wvu
+  detail::alteration_kernel<<<nblocks, nthreads, 0, stream>>>(
+    v, e, offsets, indices, weights, max,
+    thrust::raw_pointer_cast(rand_values.data()),
+    thrust::raw_pointer_cast(altered_weights.data()));
+}
+
+// updates colors of vertices by propagating the lower color to the higher
+template <typename vertex_t, typename edge_t, typename weight_t>
+void MST_solver<vertex_t, edge_t, weight_t>::label_prop(vertex_t* mst_src,
+                                                        vertex_t* mst_dst) {
+  // update the colors of both ends its until there is no change in colors
+  thrust::host_vector<edge_t> curr_mst_edge_count = mst_edge_count;
+
+  auto min_pair_nthreads = std::min(v, max_threads);
+  auto min_pair_nblocks =
+    std::min((v + min_pair_nthreads - 1) / min_pair_nthreads, max_blocks);
+
+  rmm::device_vector<bool> done(1, false);
+
+  edge_t* new_mst_edge_ptr = thrust::raw_pointer_cast(new_mst_edge.data());
+  vertex_t* color_index_ptr = thrust::raw_pointer_cast(color_index.data());
+  vertex_t* next_color_ptr = thrust::raw_pointer_cast(next_color.data());
+
+  bool* done_ptr = thrust::raw_pointer_cast(done.data());
+
+  auto i = 0;
+  while (!done[0]) {
+    done[0] = true;
+
+    detail::min_pair_colors<<<min_pair_nblocks, min_pair_nthreads, 0, stream>>>(
+      v, indices, new_mst_edge_ptr, color, color_index_ptr, next_color_ptr);
+
+    detail::update_colors<<<min_pair_nblocks, min_pair_nthreads, 0, stream>>>(
+      v, color, color_index_ptr, next_color_ptr, done_ptr);
+    i++;
+  }
+
+  detail::
+    final_color_indices<<<min_pair_nblocks, min_pair_nthreads, 0, stream>>>(
+      v, color, color_index_ptr);
+#ifdef MST_TIME
+  std::cout << "Label prop iterations: " << i << std::endl;
+#endif
+}
+
+// Finds the minimum edge from each vertex to the lowest color
+template <typename vertex_t, typename edge_t, typename weight_t>
+void MST_solver<vertex_t, edge_t, weight_t>::min_edge_per_vertex() {
+  thrust::fill(min_edge_color.begin(), min_edge_color.end(),
+               std::numeric_limits<weight_t>::max());
+
+  int n_threads = 32;
+
+  vertex_t* color_index_ptr = thrust::raw_pointer_cast(color_index.data());
```

Comment:
An observation... I have been using the `.get()` method on the pointer to retrieve this value:

```vertex_t* color_index_ptr = color_index.data().get();```

feels more pleasant to me.  I believe they are equivalent constructs.

Reply:
It is nicer indeed, thanks for the suggestion I removed all `raw_pointer_cast` in c25a559